### PR TITLE
Adding Gemmini Performance Counter

### PR DESCRIPTION
**Pull request by zitaofang**
As title. I haven't implemented the accessing instruction yet, but most of the counter design is done at this point. @hngenc  @sehuang

### CODE REVIEW

File: src/main/scala/gemmini/ExecuteController.scala [hunk 1]
```diff
@@ -6,6 +6,7 @@ import chisel3.util._
 import GemminiISA._
 import Util._
 import freechips.rocketchip.config.Parameters
+import org.w3c.dom.css.Counter
```
Comment (hngenc) [inline src/main/scala/gemmini/ExecuteController.scala:9]: I'm guessing this was an accidental import? :)

File: src/main/scala/gemmini/CounterFile.scala [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+  val RDMA_BYTE_REC = 4.U
+  val WDMA_BYTE_SENT = 5.U
+
+  val n = 6
+}
+
+class CounterEventIO extends Bundle {
+  val event_signal = Input(Vec(CounterEvent.n, Bool()))
+  val external_values = Input(Vec(CounterExternal.n, p(XLen)))
```
Comment (hngenc) [inline src/main/scala/gemmini/CounterFile.scala:90]: Is this difference between external values, and event signals that event signals can only be incremented by +1, but external values can be anything?

File: src/main/scala/gemmini/CounterFile.scala [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+// A simple counter file. Every counter is incremented when the corresponding event signal is high on rising edge.
+class CounterFile(nPerfCounter: Int, counterWidth: Int) extends Module
+{
+  val io = IO(new CounterIO(nPerfCounter, counterWidth))
+
+  val counters = Vec(nPerfCounter, RegInit(0.U(counterWidth.W), io.reset))
+  val counter_config = Vec(nPerfCounter, RegInit(0.U(CounterEvent.n)))
+  val snapshot_enable = RegInit(false.B, io.reset)
+
+  // Snapshot: In case the access instruction get blocked, it is possible to take a snapshot when reading counter
```
Comment (hngenc) [inline src/main/scala/gemmini/CounterFile.scala:151]: Why would the access instruction be "blocked"?

**Review by hngenc — COMMENTED** [review]
This looks very good, thank you. I like the new `CounterFile` classes a lot.

I posted a few questions about your implementation.

File: src/main/scala/gemmini/ExecuteController.scala (hunk 1 above)
Comment (zitaofang) [reply src/main/scala/gemmini/ExecuteController.scala:9]: Thanks for the catch! This is from my editor and I will fix it in this version.

File: src/main/scala/gemmini/CounterFile.scala (hunk 2 above)
Comment (zitaofang) [reply src/main/scala/gemmini/CounterFile.scala:90]: Exactly. Some counter will increase by more than one in this cases, so I will need to have the register placed in the module that trigger this counter and let the counter regfile manually read these external registers.

File: src/main/scala/gemmini/CounterFile.scala (hunk 3 above)
Comment (zitaofang) [reply src/main/scala/gemmini/CounterFile.scala:151]: This should be "a sequence of access instructions", I will fix the comment. If we are simulating with an OS, the thread can be preempted before we finish the sequence, making correlating the values of different counter difficult.

Comment (hngenc) [reply src/main/scala/gemmini/CounterFile.scala:151]: Ah, I see, that's good thinking

**Comment by jerryz123** [comment]
It would be useful for testing, and for regressions, to have some of the gemmini-rocc-tests use these counters.

**Comment by zitaofang** [comment]
The PR is ready.

**Comment by zitaofang** [comment]
I am closing this PR to move onto a local branch PR.

**Pull request closed without merging**